src/main.cpp

```diff
@@ -2316,8 +2326,13 @@ static bool ActivateBestChainStep(CValidationState &state, CBlockIndex *pindexMo
     }
     }
 
-    if (fBlocksDisconnected)
-        mempool.TrimToSize(GetArg("-maxmempool", DEFAULT_MAX_MEMPOOL_SIZE) * 1000000);
+    if (fBlocksDisconnected) {
+        size_t mempoolLimit = GetArg("-maxmempool", DEFAULT_MAX_MEMPOOL_SIZE) * 1000000;
+        const double defaultPriorityLimit = (double) GetArg("-blockprioritysize", DEFAULT_BLOCK_PRIORITY_SIZE) / (double) GetArg("-blockmaxsize", DEFAULT_BLOCK_MAX_SIZE);
+        size_t priorityLimit = GetArg("-prioritylimit", defaultPriorityLimit * mempoolLimit);
```

Comment:
This algorithm is done several times, and is a cluster to read.  Could you extract it out at all to make review easier?